## 모듈 import

In [1]:
from IPython.display import Image
import numpy as np
import pandas as pd
import seaborn as sns

## 데이터셋 로드

In [2]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

**컬럼(columns) 설명**

- survivied: 생존여부 (1: 생존, 0: 사망)
- pclass: 좌석 등급 (1등급, 2등급, 3등급)
- sex: 성별
- age: 나이
- sibsp: 형제 + 배우자 수
- parch: 부모 + 자녀 수
- fare: 좌석 요금
- embarked: 탑승 항구 (S, C, Q)
- class: pclass와 동일
- who: 성별과 동일
- adult_male: 성인 남자 여부
- deck: 데크 번호 (알파벳 + 숫자 혼용)
- embark_town: 탑승 항구 이름
- alive: 생존여부 (yes, no)
- alone: 혼자 탑승 여부

## 새로운 컬럼 추가

In [3]:
df1 = df.copy()

In [4]:
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

임의의 값을 **대입**하여 새로운 컬럼을 추가할 수 있습니다.

In [5]:
df1['VIP'] = True

In [6]:
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone   VIP  
0    man        True  NaN  Southampton    no  False  True  
1  woman       False    C    Cherbourg   yes  False  True  
2  woman       False  NaN  Southampton   yes   True  True  
3  woman       False    C  Southampton   yes  False  True  
4    man        True  NaN  Southampton    no   True  True

## 삭제

삭제는 **행(row) 삭제와 열(column) 삭제**로 구분할 수 있습니다.

### 행 (row) 삭제

행 삭제시 **index를 지정하여 삭제**합니다.

In [7]:
df1.drop(1)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
5           0       3    male   NaN      0      0   8.4583        Q   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone   VIP  
0      man        True  NaN  Southampton    no  False  True  
2    woman       False  NaN  Southampton   yes   True  True  
3    woman       False    C  Southampton   yes  False  True  
4      man        True  NaN  Southampton    no   True  True  
5      man        True  NaN   Queenstown    no   True  True  
..     ...         ...  ...          ...   ...    ...   ...  
886    man        True  NaN  Southampton    no   True  True  
887  woman       False    B  Southampton   yes   True  True  
888  woman       False  NaN  Southampton    no  False  True  
889    man        True    C    Cherbourg   yes   True  True  
890    man        True  NaN   Queenstown    no   True  True  

[890 rows x 16 columns]

행 삭제시 **범위를 지정하여 삭제**할 수 있습니다.

In [8]:
df1.drop(np.arange(10))

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
10          1       3  female   4.0      1      1  16.7000        S   Third   
11          1       1  female  58.0      0      0  26.5500        S   First   
12          0       3    male  20.0      0      0   8.0500        S   Third   
13          0       3    male  39.0      1      5  31.2750        S   Third   
14          0       3  female  14.0      0      0   7.8542        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone   VIP  
10   child       False    G  Southampton   yes  False  True  
11   woman       False    C  Southampton   yes   True  True  
12     man        True  NaN  Southampton    no   True  True  
13     man        True  NaN  Southampton    no  False  True  
14   child       False  NaN  Southampton    no   True  True  
..     ...         ...  ...          ...   ...    ...   ...  
886    man        True  NaN  Southampton    no   True  True  
887  woman       False    B  Southampton   yes   True  True  
888  woman       False  NaN  Southampton    no  False  True  
889    man        True    C    Cherbourg   yes   True  True  
890    man        True  NaN   Queenstown    no   True  True  

[881 rows x 16 columns]

**fancy indexing**을 활용하여 삭제할 수 있습니다.

In [9]:
df1.drop([1, 3, 5, 7, 9])

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
2           1       3  female  26.0      0      0   7.9250        S   Third   
4           0       3    male  35.0      0      0   8.0500        S   Third   
6           0       1    male  54.0      0      0  51.8625        S   First   
8           1       3  female  27.0      0      2  11.1333        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone   VIP  
0      man        True  NaN  Southampton    no  False  True  
2    woman       False  NaN  Southampton   yes   True  True  
4      man        True  NaN  Southampton    no   True  True  
6      man        True    E  Southampton    no   True  True  
8    woman       False  NaN  Southampton   yes  False  True  
..     ...         ...  ...          ...   ...    ...   ...  
886    man        True  NaN  Southampton    no   True  True  
887  woman       False    B  Southampton   yes   True  True  
888  woman       False  NaN  Southampton    no  False  True  
889    man        True    C    Cherbourg   yes   True  True  
890    man        True  NaN   Queenstown    no   True  True  

[886 rows x 16 columns]

### 열 (column) 삭제

In [10]:
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone   VIP  
0    man        True  NaN  Southampton    no  False  True  
1  woman       False    C    Cherbourg   yes  False  True  
2  woman       False  NaN  Southampton   yes   True  True  
3  woman       False    C  Southampton   yes  False  True  
4    man        True  NaN  Southampton    no   True  True

열 삭제시 **반드시 `axis=1` 옵션을 지정**해야 합니다.

In [13]:
df1.drop('class', axis=1).head()

survived  pclass     sex   age  sibsp  parch     fare embarked    who  \
0         0       3    male  22.0      1      0   7.2500        S    man   
1         1       1  female  38.0      1      0  71.2833        C  woman   
2         1       3  female  26.0      0      0   7.9250        S  woman   
3         1       1  female  35.0      1      0  53.1000        S  woman   
4         0       3    male  35.0      0      0   8.0500        S    man   

   adult_male deck  embark_town alive  alone   VIP  
0        True  NaN  Southampton    no  False  True  
1       False    C    Cherbourg   yes  False  True  
2       False  NaN  Southampton   yes   True  True  
3       False    C  Southampton   yes  False  True  
4        True  NaN  Southampton    no   True  True

**다수의 컬럼(column) 삭제**도 가능합니다.

In [ ]:
df1.drop(['who', 'deck', 'alive'], axis=1)

삭제된 내용을 바로 적용하려면 `inplace=True`를 지정합니다.

In [14]:
df1.drop(['who', 'deck', 'alive'], axis=1, inplace=True)

In [15]:
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

   adult_male  embark_town  alone   VIP  
0        True  Southampton  False  True  
1       False    Cherbourg  False  True  
2       False  Southampton   True  True  
3       False  Southampton  False  True  
4        True  Southampton   True  True

## 연습문제

- `df1`에서 1, 3, 5번행을 삭제해 주세요
- `df1`에서 `embarked`, `class`, `alone` 컬럼을 삭제해 주세요
- `df1`의 상위 10개 행을 출력하세요

In [20]:
df1 = df.copy()

In [21]:
# 코드를 입력해 주세요
df1.drop([1, 3, 5], inplace=True)
# df1 = df1.drop([1, 3, 5])
df1.drop(['embarked', 'class', 'alone'], axis=1, inplace=True)
df1.head(10)

survived  pclass     sex   age  sibsp  parch     fare    who  adult_male  \
0          0       3    male  22.0      1      0   7.2500    man        True   
2          1       3  female  26.0      0      0   7.9250  woman       False   
4          0       3    male  35.0      0      0   8.0500    man        True   
6          0       1    male  54.0      0      0  51.8625    man        True   
7          0       3    male   2.0      3      1  21.0750  child       False   
8          1       3  female  27.0      0      2  11.1333  woman       False   
9          1       2  female  14.0      1      0  30.0708  child       False   
10         1       3  female   4.0      1      1  16.7000  child       False   
11         1       1  female  58.0      0      0  26.5500  woman       False   
12         0       3    male  20.0      0      0   8.0500    man        True   

   deck  embark_town alive  
0   NaN  Southampton    no  
2   NaN  Southampton   yes  
4   NaN  Southampton    no  
6     E  Southampton    no  
7   NaN  Southampton    no  
8   NaN  Southampton   yes  
9   NaN    Cherbourg   yes  
10    G  Southampton   yes  
11    C  Southampton   yes  
12  NaN  Southampton    no

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>survived</th>
      <th>pclass</th>
      <th>sex</th>
      <th>age</th>
      <th>sibsp</th>
      <th>parch</th>
      <th>fare</th>
      <th>who</th>
      <th>adult_male</th>
      <th>deck</th>
      <th>embark_town</th>
      <th>alive</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>3</td>
      <td>male</td>
      <td>22.0</td>
      <td>1</td>
      <td>0</td>
      <td>7.2500</td>
      <td>man</td>
      <td>True</td>
      <td>NaN</td>
      <td>Southampton</td>
      <td>no</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1</td>
      <td>3</td>
      <td>female</td>
      <td>26.0</td>
      <td>0</td>
      <td>0</td>
      <td>7.9250</td>
      <td>woman</td>
      <td>False</td>
      <td>NaN</td>
      <td>Southampton</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>3</td>
      <td>male</td>
      <td>35.0</td>
      <td>0</td>
      <td>0</td>
      <td>8.0500</td>
      <td>man</td>
      <td>True</td>
      <td>NaN</td>
      <td>Southampton</td>
      <td>no</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0</td>
      <td>1</td>
      <td>male</td>
      <td>54.0</td>
      <td>0</td>
      <td>0</td>
      <td>51.8625</td>
      <td>man</td>
      <td>True</td>
      <td>E</td>
      <td>Southampton</td>
      <td>no</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0</td>
      <td>3</td>
      <td>male</td>
      <td>2.0</td>
      <td>3</td>
      <td>1</td>
      <td>21.0750</td>
      <td>child</td>
      <td>False</td>
      <td>NaN</td>
      <td>Southampton</td>
      <td>no</td>
    </tr>
    <tr>
      <th>8</th>
      <td>1</td>
      <td>3</td>
      <td>female</td>
      <td>27.0</td>
      <td>0</td>
      <td>2</td>
      <td>11.1333</td>
      <td>woman</td>
      <td>False</td>
      <td>NaN</td>
      <td>Southampton</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>9</th>
      <td>1</td>
      <td>2</td>
      <td>female</td>
      <td>14.0</td>
      <td>1</td>
      <td>0</td>
      <td>30.0708</td>
      <td>child</td>
      <td>False</td>
      <td>NaN</td>
      <td>Cherbourg</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>10</th>
      <td>1</td>
      <td>3</td>
      <td>female</td>
      <td>4.0</td>
      <td>1</td>
      <td>1</td>
      <td>16.7000</td>
      <td>child</td>
      <td>False</td>
      <td>G</td>
      <td>Southampton</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>11</th>
      <td>1</td>
      <td>1</td>
      <td>female</td>
      <td>58.0</td>
      <td>0</td>
      <td>0</td>
      <td>26.5500</td>
      <td>woman</td>
      <td>False</td>
      <td>C</td>
      <td>Southampton</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>12</th>
      <td>0</td>
      <td>3</td>
      <td>male</td>
      <td>20.0</td>
      <td>0</td>
      <td>0</td>
      <td>8.0500</td>
      <td>man</td>
      <td>True</td>
      <td>NaN</td>
      <td>Southampton</td>
      <td>no</td>
    </tr>
  </tbody>
</table>
</div>

## 컬럼간 연산

**컬럼(column) 과 컬럼 사이의 연산을 매우 쉽게 적용**할 수 있습니다.

In [22]:
df1 = df.copy()

**family(가족)**의 총합은 **sibsp**컬럼과 **parch**의 합산으로 구할 수 있습니다.

In [24]:
df1['family'] = df1['sibsp'] + df1['parch']

In [25]:
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  family  
0    man        True  NaN  Southampton    no  False       1  
1  woman       False    C    Cherbourg   yes  False       1  
2  woman       False  NaN  Southampton   yes   True       0  
3  woman       False    C  Southampton   yes  False       1  
4    man        True  NaN  Southampton    no   True       0

**문자열의 합 (이어붙히기)도 가능**합니다.

In [26]:
df1['gender'] = df1['who'] + '-' + df1['sex']

In [27]:
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  family        gender  
0    man        True  NaN  Southampton    no  False       1      man-male  
1  woman       False    C    Cherbourg   yes  False       1  woman-female  
2  woman       False  NaN  Southampton   yes   True       0  woman-female  
3  woman       False    C  Southampton   yes  False       1  woman-female  
4    man        True  NaN  Southampton    no   True       0      man-male

컬럼간 연산시 `round()`를 사용하여 소수점 자릿수를 지정할 수 있습니다.

**round(숫자, 소수 몇 째자리)**

In [28]:
df1['round'] = round(df1['fare'] / df1['age'], 2)

In [29]:
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  family        gender  \
0    man        True  NaN  Southampton    no  False       1      man-male   
1  woman       False    C    Cherbourg   yes  False       1  woman-female   
2  woman       False  NaN  Southampton   yes   True       0  woman-female   
3  woman       False    C  Southampton   yes  False       1  woman-female   
4    man        True  NaN  Southampton    no   True       0      man-male   

   round  
0   0.33  
1   1.88  
2   0.30  
3   1.52  
4   0.23

연산시 1개의 컬럼이라도 **NaN 값을 포함하고 있다면 결과는 NaN** 이 됩니다.

In [30]:
df1.loc[df1['age'].isnull(), 'deck':].head()

deck  embark_town alive  alone  family        gender  round
5   NaN   Queenstown    no   True       0      man-male    NaN
17  NaN  Southampton   yes   True       0      man-male    NaN
19  NaN    Cherbourg   yes   True       0  woman-female    NaN
26  NaN    Cherbourg    no   True       0      man-male    NaN
28  NaN   Queenstown   yes   True       0  woman-female    NaN

## 타입 변환 (astype)

In [31]:
df1 = df.copy()

In [32]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


`int32`로 변경

In [33]:
df1['pclass'].astype('int32').head()

0    3
1    1
2    3
3    1
4    3
Name: pclass, dtype: int32

`float32`로 변경

In [34]:
df1['pclass'].astype('float32').head()

0    3.0
1    1.0
2    3.0
3    1.0
4    3.0
Name: pclass, dtype: float32

`object`로 변경

In [ ]:
df1['pclass'].astype('str').head()

`category`로 변경.

`category`로 변경시에는 Categories가 같이 출력 됩니다.

In [35]:
df1['who'].value_counts()

who
man      537
woman    271
child     83
Name: count, dtype: int64

In [36]:
df1['who'].dtype

dtype('O')

In [37]:
df1['who'].astype('category').head()

0      man
1    woman
2    woman
3    woman
4      man
Name: who, dtype: category
Categories (3, object): ['child', 'man', 'woman']

타입을 `category`로 변환했다면 **.cat**으로 접근하여 category 타입이 제공하는 **attribute를 사용**할 수 있습니다.

In [38]:
df1['who'] = df1['who'].astype('category')

In [39]:
df1['who'].dtype

CategoricalDtype(categories=['child', 'man', 'woman'], ordered=False, categories_dtype=object)

In [40]:
df1['who'].cat.codes

0      1
1      2
2      2
3      2
4      1
      ..
886    1
887    2
888    2
889    1
890    1
Length: 891, dtype: int8

**카테고리 이름 변경**

In [42]:
# df1['who'].cat.categories
[f'Group({g})' for g in df1['who'].cat.categories]

['Group(child)', 'Group(man)', 'Group(woman)']

In [43]:
df1['who'] = df1['who'].cat.rename_categories([f'Group({g})' for g in df1['who'].cat.categories])
df1['who'].value_counts()

who
Group(man)      537
Group(woman)    271
Group(child)     83
Name: count, dtype: int64

## datetime - 날짜, 시간

### date_range

주요 옵션 값
- **start**: 시작 날짜
- **end**: 끝 날짜
- **periods**: 생성할 데이터 개수
- **freq**: 주기

In [46]:
dates = pd.date_range('20210101', periods=df.shape[0], freq='15H')
dates
dates.shape

(891,)

In [47]:
df1 = df.copy()
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

**date의 컬럼**을 만들어 생성한 **date 를 대입**합니다.

In [48]:
df1['date'] = dates
df1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone                date  
0    man        True  NaN  Southampton    no  False 2021-01-01 00:00:00  
1  woman       False    C    Cherbourg   yes  False 2021-01-01 15:00:00  
2  woman       False  NaN  Southampton   yes   True 2021-01-02 06:00:00  
3  woman       False    C  Southampton   yes  False 2021-01-02 21:00:00  
4    man        True  NaN  Southampton    no   True 2021-01-03 12:00:00

In [49]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   survived     891 non-null    int64         
 1   pclass       891 non-null    int64         
 2   sex          891 non-null    object        
 3   age          714 non-null    float64       
 4   sibsp        891 non-null    int64         
 5   parch        891 non-null    int64         
 6   fare         891 non-null    float64       
 7   embarked     889 non-null    object        
 8   class        891 non-null    category      
 9   who          891 non-null    object        
 10  adult_male   891 non-null    bool          
 11  deck         203 non-null    category      
 12  embark_town  889 non-null    object        
 13  alive        891 non-null    object        
 14  alone        891 non-null    bool          
 15  date         891 non-null    datetime64[ns]
dtypes: bool(

**date**의 컬럼에 `datetime64`라는 데이터 타입이 표기됩니다.

### datetime 타입

`datetime` 타입에서는 **dt** 접근자로 다음과 같은 날짜 속성에 쉽게 접근할 수 있습니다.

Pandas의 **dt (datetime) 날짜 관련 변수**는 다음과 같습니다.

[도큐먼트](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.year.html)

- pandas.Series.dt.year: 연도
- pandas.Series.dt.month: 월
- pandas.Series.dt.day: 일
- pandas.Series.dt.hour: 시
- pandas.Series.dt.minute: 분
- pandas.Series.dt.second: 초
- pandas.Series.dt.microsecond: micro 초
- pandas.Series.dt.nanosecond: nano 초
- pandas.Series.dt.week: 주
- pandas.Series.dt.weekofyear: 연중 몇 째주
- pandas.Series.dt.dayofweek: 요일
- pandas.Series.dt.weekday: 요일 (dayofweek과 동일)
- pandas.Series.dt.dayofyear: 연중 몇 번째 날
- pandas.Series.dt.quarter: 분기

In [50]:
# 연도
df1['date'].dt.year.head()

0    2021
1    2021
2    2021
3    2021
4    2021
Name: date, dtype: int32

In [51]:
# 월
df1['date'].dt.month.head()

0    1
1    1
2    1
3    1
4    1
Name: date, dtype: int32

In [52]:
# 일
df1['date'].dt.day.head()

0    1
1    1
2    2
3    2
4    3
Name: date, dtype: int32

**dayofweek**는 숫자로 요일이 표기 됩니다.
- 월요일: 0, 일요일: 6

In [53]:
df1['date'].dt.dayofweek.head(10)

0    4
1    4
2    5
3    5
4    6
5    0
6    0
7    1
8    2
9    2
Name: date, dtype: int32

### to_datetime

In [58]:
# e notation 표현 방식 변경
pd.options.display.float_format = '{:.2f}'.format
df1['date']

0     2021-01-01 00:00:00
1     2021-01-01 15:00:00
2     2021-01-02 06:00:00
3     2021-01-02 21:00:00
4     2021-01-03 12:00:00
              ...        
886   2022-07-08 18:00:00
887   2022-07-09 09:00:00
888   2022-07-10 00:00:00
889   2022-07-10 15:00:00
890   2022-07-11 06:00:00
Name: date, Length: 891, dtype: datetime64[ns]

샘플용 **서울시 공공자전거 데이터를 로드**합니다.

In [81]:
!pip install teddynote -q

from teddynote import dataset

dataset.download('서울시자전거')

^C
======= 다운로드 시작 =======




[notice] A new release of pip is available: 23.2.1 -> 23.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


data\seoul_bicycle.csv


ImportError: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html

In [ ]:
import pandas as pd

In [ ]:
# 데이터셋 로드
df2 = pd.read_csv('data/seoul_bicycle.csv')
df2.head()

In [ ]:
df2.info()

**대여일자** 컬럼은 날짜 관련 컬럼처럼 보이나 `info()`는 object로 인식하였습니다.

`datetime`타입으로 변경해야 .dt 접근자를 사용할 수 있습니다.

**`pd.to_datetime()`**: datetime type으로 변환합니다.

In [ ]:
pd.to_datetime(df2['대여일자'])

재대입하여 **컬럼에 적용**합니다.

In [ ]:
df2['대여일자'] = pd.to_datetime(df2['대여일자'])

In [ ]:
df2.info()

적용된 후 `.dt`접근자를 활용하여 datetime 속성에 접근할 수 있습니다.

In [ ]:
df2['대여일자'].dt.year

In [ ]:
df2['대여일자'].dt.month

In [ ]:
df2['대여일자'].dt.day

In [ ]:
df2['대여일자'].dt.dayofweek

## pd.to_numeric() - 수치형 변환

object나 numerical type이 아닌 컬럼을 **수치형(numerical) 컬럼으로 변환**할 때 사용합니다.

In [ ]:
df2.head()

In [ ]:
df2.info()

**운동량** 컬럼은 숫자형 컬럼 처럼 보이지만, **object 타입으로 지정**되어 있습니다. 

종종 이런 현상이 발생하는데, 이런 현상을 만들어낸 이유는 분명 존재합니다!

원인 파악을 위해서 일단 `pd.to_numeric()`으로 **변환을 시도**합니다.

In [ ]:
pd.to_numeric(df2['운동량'])

**2344 position**에 무언가 에러가 발생하였습니다.

In [ ]:
df2.loc[2344]

운동량에 숫자형이 아닌 개행 (\N)이 들어가 있기 때문에 이러한 에러가 발생하였습니다.

숫자형으로 바꿀 때 **NaN값이나 숫자로 변환이 불가능한 문자열이 존재할 때 변환에 실패**하게 됩니다.

`errors=` 옵션 값을 바꾸어 해결할 수 있습니다.

errors : {'ignore', 'raise', 'coerce'}, default 'raise'
- If 'raise', then invalid parsing will raise an exception
- If 'coerce', then invalid parsing will be set as NaN
- If 'ignore', then invalid parsing will return the input

`errors='coerce'`로 지정하면 잘못된 문자열은 **NaN 값으로 치환하여 변환**합니다. 

그리고, 결과 확인시 잘 변환이 된 것을 볼 수 있습니다.

In [ ]:
pd.to_numeric(df2['운동량'], errors='coerce')

In [ ]:
pd.to_numeric(df2['운동량'], errors='coerce').loc[2344]

`errors='ignore'`로 지정하게 되면 잘못된 문자열이 숫자로 **변환이 안되고 무시**하기 때문에 전체 컬럼의 dtype이 **object로 그대로 남아있습니다.**

In [ ]:
pd.to_numeric(df2['운동량'], errors='ignore')

In [ ]:
pd.to_numeric(df2['운동량'], errors='ignore').loc[2344]

재대입까지 마무리 해야 DataFrame에 적용됩니다.

In [ ]:
df2['운동량'] = pd.to_numeric(df2['운동량'], errors='coerce')

In [ ]:
df2.info()

## pd.cut() - 구간 나누기(binning)

연속된 수치(continuous values)를 **구간으로 나누어 카테고리화** 할 때 사용합니다.

In [ ]:
df2.head()

직접 범위 설정을 해줄 수 있습니다.
- `right=False`로 지정시 우측 범위를 포함하지 않습니다.

In [ ]:
bins = [0, 200, 400, df2['운동량'].max()]
pd.cut(df2['운동량'], bins, right=False)

`labels`를 지정해 줄 수 있으며, 지정한 bins의 개수보다 1 개가 적어야 합니다.

In [ ]:
labels = ['운동부족', '보통', '많음']

In [ ]:
pd.cut(df2['운동량'], bins, labels=labels, right=False)

In [ ]:
df2['운동량'].describe()

운동량은 범위가 굉장히 넓습니다. **최소값은 0인데, 최대값은 엄청 큰 값**이 존재합니다.

어쨌든, 운동량을 기준으로 데이터를 10개 그룹으로 분류하고 싶습니다.

`pd.cut()`을 활용하여 쉽게 그룹을 나눌 수 있습니다.

In [80]:
df2['운동량'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 327231 entries, 0 to 327230
Series name: 운동량
Non-Null Count   Dtype 
--------------   ----- 
327231 non-null  object
dtypes: object(1)
memory usage: 2.5+ MB


`bins` 옵션에 나누고자 하는 **구간의 개수**를 설정합니다.

In [78]:
df2['운동량'] = df2['운동량'].fillna(df2['운동량'].median())
# df2['운동량'] = df2['운동량'].astype('int64')
df2['운동량_cut'] = pd.cut(df2['운동량'], 10)

TypeError: Cannot convert ['61.82' '39.62' '430.85' ... '19.96' '43.77' '4735.63'] to numeric

In [ ]:
df2['운동량_cut'].value_counts()

분포를 보니 첫 구간에 대부분의 데이터가 쏠려 있습니다. 딱봐도 올바르지 않은 방법 같아 보입니다.

`pd.cut()`은 **최소에서 최대 구간을 지정한 bin만큼 동일하게 분할** 하기 때문에 이런 현상이 발생할 수 있습니다.

고르게 분포한 데이터라면 괜찮지만, 튀는 **이상치(outlier)가 있는 경우에는 안 좋은 결과**를 초래 합니다.

## pd.qcut() - 동일한 개수를 갖도록 구간 분할

`pd.cut()`과 유사하지만, **quantity 즉 데이터의 분포를 최대한 비슷하게 유지**하는 구간을 분할 합니다.

In [112]:
df2['운동량_qcut'] = pd.qcut(df2['운동량'], q=10)

In [113]:
df2['운동량_qcut'].value_counts()

운동량_qcut
(93.414, 192.02]           32690
(-0.001, 24.737]           32683
(24.737, 93.414]           32683
(601.705, 1079.744]        32683
(1079.744, 1889.606]       32683
(1889.606, 3328.186]       32683
(3328.186, 6805.188]       32683
(6805.188, 163936052.3]    32683
(344.45, 601.705]          32680
(192.02, 344.45]           32679
Name: count, dtype: int64

구간도 예쁘게 분할(**균등하게 분할**)이 된 것 처럼 보입니다. 하지만, **간격은 일정하지 않습니다.**

qcut 또한 임의 범위를 조정할 수 있습니다.

In [114]:
qcut_bins = [0, 0.2, 0.8, 1]

In [115]:
pd.qcut(df2['운동량'], qcut_bins)

0                (-0.001, 93.414]
1                (-0.001, 93.414]
2              (93.414, 3328.186]
3                (-0.001, 93.414]
4         (3328.186, 163936052.3]
                   ...           
327226         (93.414, 3328.186]
327227           (-0.001, 93.414]
327228           (-0.001, 93.414]
327229           (-0.001, 93.414]
327230    (3328.186, 163936052.3]
Name: 운동량, Length: 327231, dtype: category
Categories (3, interval[float64, right]): [(-0.001, 93.414] < (93.414, 3328.186] < (3328.186, 163936052.3]]

qcut 역시 label을 지정할 수 있습니다. 마찬가지로 범위 수보다 1개 적게 설정합니다.

In [116]:
qcut_labels = ['적음', '보통', '많음']

In [117]:
pd.qcut(df2['운동량'], qcut_bins, labels=qcut_labels).value_counts()

운동량
보통    196098
적음     65366
많음     65366
Name: count, dtype: int64